### PR TITLE
Core: reorganize delete_replicas #1431 #4221

### PR DESCRIPTION
**Pull request by rcarpa**
Refactor the delete_replicas function
- reorganize the very long lines of queries to make them easier to read
- add comments
- split the function in two

### CODE REVIEW

File: lib/rucio/core/replica.py [hunk 1]
```diff
@@ -1462,63 +1463,108 @@ def delete_replicas(rse_id, files, ignore_availability=True, session=None):
...
         raise exception.ReplicaNotFound("One or several replicas don't exist.")
 
+    __update_metadata_after_replica_deletion(rse_id=rse_id, files=files, session=session)
+
+    # Decrease RSE counter
+    decrease(rse_id=rse_id, files=delta, bytes=bytes, session=session)
+
+
+@transactional_session
+def __update_metadata_after_replica_deletion(rse_id, files, session=None):
```
Comment (bari12) [inline lib/rucio/core/replica.py]: I would call this `__cleanup_after_replica_deletion`
`metadata` has a bit of a double-loaded meaning in Rucio. I think cleanup is cleaner 😄

Comment (rcarpa) [reply lib/rucio/core/replica.py]: Done

File: lib/rucio/core/replica.py [hunk 2]
```diff
@@ -1544,21 +1590,47 @@ def delete_replicas(rse_id, files, ignore_availability=True, session=None):
...
+                # (it was the last children), metadata cleanup has to be done:
+                #
+                # 1) Schedule to remove the replicas of this empty collection
+                clt_replica_condition.append(
+                    and_(models.CollectionReplica.scope == parent_scope,
+                         models.CollectionReplica.name == parent_name,
+                         exists(select([1]).prefix_with("/*+ INDEX(DIDS DIDS_PK) */", dialect='oracle')).where(
+                             and_(models.DataIdentifier.scope == parent_scope,
+                                  models.DataIdentifier.name == parent_name,
+                                  models.DataIdentifier.is_open == False)),  # NOQA
```
Comment (cserf) [inline lib/rucio/core/replica.py:1610]: Does anyone remember why we have this ? I suspect this is the cause of this bug: https://github.com/rucio/rucio/issues/3311

**Review by cserf — COMMENTED** [review]
That looks good, but I think I found a bug (not related to your code, but in the previous code)

File: lib/rucio/core/replica.py (hunk 2 above)
Comment (bari12) [reply lib/rucio/core/replica.py:1610]: I can't remember, but thinking about this, it is probably wrong and should not filter on `is_open == False`.

Comment (bari12) [reply lib/rucio/core/replica.py:1610]: Let's continue discussing this in #3311
I think @rcarpa can have a look at this one too during his reaper2 works.

**Review by bari12 — APPROVED** [review]

**Pull request merged**